Library crashes when compiled with _GLIBCXX_DEBUG

In our project we use [_GLIBCXX_DEBUG](https://gcc.gnu.org/onlinedocs/libstdc++/manual/debug_mode.html) macro definition to enable additional safety checking in standard library.
But Botan and its unit tests compiled with this macro constantly crash.

Typical error (produced by botan-test):
```
/usr/include/c++/4.9/debug/vector:357:error: attempt to subscript container 
    with out-of-bounds index 140, but container only holds 140 elements.

Objects involved in the operation:
sequence "this" @ 0x0x7fff62ec6c90 {
  type = NSt7__debug6vectorIjN5Botan16secure_allocatorIjEEEE;
}
```

The problem is in widely-used among the code pattern:
```
algorithm(&vector[offset], length)
```

When offset >= vector.size(), operator[] will raise an error even if length == 0.
Since _GLIBCXX_DEBUG macro is intended to increase code stability and safety, I think that supporting it in cryptographic library like Botan would be beneficial.

@webmaster128,

Still fails on TLS test:
```
$ ./botan-test tls    
/usr/include/c++/4.9/bits/stl_algo.h:5224:error: function requires a valid 
    iterator range [__first2, __last2).

Objects involved in the operation:
iterator "__first2" @ 0x0x7ffe0896ba70 {
type = N11__gnu_debug14_Safe_iteratorISt23_Rb_tree_const_iteratorIN5Botan3TLS24Handshake_Extension_TypeEENSt7__debug3setIS4_St4lessIS4_ESaIS4_EEEEE (mutable iterator);
  state = dereferenceable (start-of-sequence);
  references sequence with type `NSt7__debug3setIN5Botan3TLS24Handshake_Extension_TypeESt4lessIS3_ESaIS3_EEE' @ 0x0x7ffe0896baf0
}
iterator "__last2" @ 0x0x7ffe0896bb90 {
type = N11__gnu_debug14_Safe_iteratorISt23_Rb_tree_const_iteratorIN5Botan3TLS24Handshake_Extension_TypeEENSt7__debug3setIS4_St4lessIS4_ESaIS4_EEEEE (mutable iterator);
  state = past-the-end;
  references sequence with type `NSt7__debug3setIN5Botan3TLS24Handshake_Extension_TypeESt4lessIS3_ESaIS3_EEE' @ 0x0x7ffe0896bb40
}
Aborted (core dumped)
```

Backtrace:
```
0x00007ffff6df1267 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:55
55      ../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  0x00007ffff6df1267 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:55
#1  0x00007ffff6df2eca in __GI_abort () at abort.c:89
#2  0x00007ffff797a845 in __gnu_debug::_Error_formatter::_M_error() const () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x0000000000641766 in std::set_difference<__gnu_debug::_Safe_iterator<std::_Rb_tree_const_iterator<Botan::TLS::Handshake_Extension_Type>, std::__debug::set<Botan::TLS::Handshake_Extension_Type, std::less<Botan::TLS::Handshake_Extension_Type>, std::allocator<Botan::TLS::Handshake_Extension_Type> > >, __gnu_debug::_Safe_iterator<std::_Rb_tree_const_iterator<Botan::TLS::Handshake_Extension_Type>, std::__debug::set<Botan::TLS::Handshake_Extension_Type, std::less<Botan::TLS::Handshake_Extension_Type>, std::allocator<Botan::TLS::Handshake_Extension_Type> > >, std::back_insert_iterator<std::__debug::vector<Botan::TLS::Handshake_Extension_Type, std::allocator<Botan::TLS::Handshake_Extension_Type> > > >
    (__first1=Botan::TLS::TLSEXT_ALPN, __last1=Botan::TLS::TLSEXT_CLIENT_CERT_URL, __first2=Botan::TLS::TLSEXT_SERVER_NAME_INDICATION, __last2=Botan::TLS::TLSEXT_CLIENT_CERT_URL, __result=...)
    at /usr/include/c++/4.9/bits/stl_algo.h:5224
#4  0x0000000000599197 in Botan::TLS::Client::process_handshake_msg (this=0x7fffffffd300, active_state=0x0, state_base=..., type=Botan::TLS::SERVER_HELLO, contents=std::__debug::vector of length 90, capacity 90 = {...})
    at botan_all.cpp:43590
#5  0x0000000000594f2d in Botan::TLS::Channel::received_data (this=0x7fffffffd300, input=0xb6e053 "\026\003\001\004D\v", input_size=1428) at botan_all.cpp:42776
#6  0x000000000043f92b in (anonymous namespace)::basic_test_handshake (rng=..., offer_version=..., creds=..., policy=...) at ./src/tests/unit_tls.cpp:248
#7  0x000000000043ff2b in test_tls () at ./src/tests/unit_tls.cpp:298
```

But all other tests are seems to be fine.

Yes, 170 works and now, as far as I can see, everything compiles and runs fine.

I also checked tried compiling with Clang:
With -D_GLIBCPP_DEBUG it generates the following errors:
```
/usr/include/c++/v1/algorithm:1771:12: error: no viable conversion from 'typename enable_if<is_same<typename remove_const<const unsigned int>::type, unsigned int>::value && is_trivially_copy_assignable<unsigned int>::value, unsigned int
      *>::type' (aka 'unsigned int *') to 'std::__1::__wrap_iter<unsigned int *>'
    return _VSTD::__copy(__unwrap_iter(__first), __unwrap_iter(__last), __unwrap_iter(__result));
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/include/c++/v1/__config:358:15: note: expanded from macro '_VSTD'
#define _VSTD std::_LIBCPP_NAMESPACE
              ^
botan_all.cpp:5725:9: note: in instantiation of function template specialization 'std::__1::copy<const unsigned int *, std::__1::__wrap_iter<unsigned int *> >' requested here
   std::copy(P_INIT, P_INIT + 18, P.begin());
        ^
/usr/include/c++/v1/iterator:1140:5: note: candidate constructor not viable: no known conversion from 'typename enable_if<is_same<typename remove_const<const unsigned int>::type, unsigned int>::value &&
      is_trivially_copy_assignable<unsigned int>::value, unsigned int *>::type' (aka 'unsigned int *') to 'const std::__1::__wrap_iter<unsigned int *> &' for 1st argument
    __wrap_iter(const __wrap_iter& __x)
    ^
/usr/include/c++/v1/iterator:1130:52: note: candidate template ignored: could not match '__wrap_iter<type-parameter-0-0>' against 'unsigned int *'
    template <class _Up> _LIBCPP_INLINE_VISIBILITY __wrap_iter(const __wrap_iter<_Up>& __u,
                                                   ^
67 warnings and 1 error generated.
Makefile:131: recipe for target 'build/obj/lib/botan_all.o' failed
make: *** [build/obj/lib/botan_all.o] Error 1
```

Which can be fixed by this patch:
```
--- Botan-1.11.18/src/lib/block/blowfish/blowfish.cpp   2015-07-06 04:20:48.000000000 +0300
+++ Botan-1.11.18/src/lib/block/blowfish/blowfish.cpp   2015-07-06 21:58:42.043124065 +0300
@@ -88,10 +88,10 @@
 void Blowfish::key_schedule(const byte key[], size_t length)
    {
    P.resize(18);
-   std::copy(P_INIT, P_INIT + 18, P.begin());
+   std::memcpy(&P[0], P_INIT, sizeof(P_INIT));

    S.resize(1024);
-   std::copy(S_INIT, S_INIT + 1024, S.begin());
+   std::memcpy(&S[0], S_INIT, sizeof(S_INIT));

    const byte null_salt[16] = { 0 };

@@ -133,10 +133,10 @@
                                   std::to_string(workfactor) + " too large");

    P.resize(18);
-   std::copy(P_INIT, P_INIT + 18, P.begin());
+   std::memcpy(&P[0], P_INIT, sizeof(P_INIT));

    S.resize(1024);
-   std::copy(S_INIT, S_INIT + 1024, S.begin());
+   std::memcpy(&S[0], S_INIT, sizeof(S_INIT));

    key_expansion(key, length, salt);
```

With -D_LIBCPP_DEBUG=0 (headers only) it compiles and works OOTB. 

Thank you.

--- a/src/lib/asn1/asn1_obj.cpp
+++ b/src/lib/asn1/asn1_obj.cpp
@@ -10,6 +10,7 @@
 #include <botan/ber_dec.h>
 #include <botan/data_src.h>
 #include <botan/parsing.h>
+#include <botan/internal/stl_util.h>
 
 namespace Botan {
 
@@ -45,8 +46,7 @@ std::vector<byte> put_in_sequence(const std::vector<byte>& contents)
 */
 std::string to_string(const BER_Object& obj)
    {
-   return std::string(reinterpret_cast<const char*>(&obj.value[0]),
-                      obj.value.size());
+   return to_string(obj.value);
    }
 
 /*

--- a/src/lib/utils/stl_util.h
+++ b/src/lib/utils/stl_util.h
@@ -1,6 +1,7 @@
 /*
 * STL Utility Functions
 * (C) 1999-2007 Jack Lloyd
+* (C) 2015 Simon Warta (Kullo GmbH)
 *
 * Botan is released under the Simplified BSD License (see license.txt)
 */
@@ -11,12 +12,18 @@
 #include <vector>
 #include <string>
 #include <map>
+#include <botan/secmem.h>
 
 namespace Botan {
 
 inline std::vector<byte> to_byte_vector(const std::string& s)
    {
    return std::vector<byte>(s.cbegin(), s.cend());
+   }
+
+inline std::string to_string(const secure_vector<byte> &bytes)
+   {
+   return std::string(bytes.cbegin(), bytes.cend());
    }
 
 /*

